### PR TITLE
fix: Don't bypass other tc-bpf programs

### PR DESCRIPTION
**Pull request by pchaigno**
# Description

This pull request fixes the compatibility with Cilium. Both Cilium and Retina attach tc-bpf programs to the host interfaces of pods and, depending on the configuration for Cilium, to the native devices (eth0 & co). As a result, whichever program is executed first bypasses the other program. Since Cilium relies on its BPF programs to implement service handling, policy enforcement, and much more, a bypass can have dire consequences.

To avoid bypassing Cilium, Retina needs to return TC_ACT_UNSPEC instead of TC_ACT_OK; that return code will cause subsequent BPF programs attach at the same tc hook to be executed [1]. In addition, Cilium should be run with `--bpf-filter-priority=2` to always run after Retina.

This change was tested on a 5-nodes GKE cluster with Cilium installed. The full connectivity tests from latest [Cilium CLI](https://github.com/cilium/cilium-cli/blob/main/README.md#connectivity-check) passed. Retina was attached to both the ingress and egress of all interfaces:

    $ ks exec ds/cilium -c cilium-agent -- bpftool net | grep lxc461b2341ecd6
    lxc461b2341ecd6(18) clsact/ingress endpoint_ingres id 3283
    lxc461b2341ecd6(18) clsact/ingress cil_from_container-lxc461b2341ecd6 id 790
    lxc461b2341ecd6(18) clsact/egress endpoint_egress id 3282
    lxc461b2341ecd6(18) clsact/egress cil_to_container-lxc461b2341ecd6 id 795

1 - https://lpc.events/event/16/contributions/1353/attachments/1068/2130/plumbers_2022_tc_bpf_links.pdf

## Related Issue

https://github.com/microsoft/retina/issues/252

## Checklist

- [x] I have read the [contributing documentation](https://retina.sh/docs/contributing).
- [x] I signed and signed-off the commits (`git commit -S -s ...`). See [this documentation](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) on signing commits.
- [x] I have correctly attributed the author(s) of the code.
- [x] I have tested the changes locally.
- [x] I have followed the project's style guidelines.
- [x] I have updated the documentation, if necessary.
- [ ] I have added tests, if applicable.

## Screenshots (if applicable) or Testing Completed

Please add any relevant screenshots or GIFs to showcase the changes made.

## Additional Notes

Add any additional notes or context about the pull request here.

---

Please refer to the [CONTRIBUTING.md](../CONTRIBUTING.md) file for more information on how to contribute to this project.

### CODE REVIEW

File: pkg/plugin/packetparser/_cprog/packetparser.c [hunk 1]
```diff
@@ -237,7 +237,7 @@ int endpoint_ingress_filter(struct __sk_buff *skb)
 	// So ingress on host is egress on endpoint and vice versa.
 	parse(skb, FROM_ENDPOINT);
 	// Always return 0 to allow packet to pass.
-	return 0;
+	return -1;
```
Comment (tamilmani1989) [inline pkg/plugin/packetparser/_cprog/packetparser.c:240]: can we include `linux/pkt_cls.h` and return `TC_ACT_UNSPEC` instead of -1?

**Review by anubhabMajumdar — DISMISSED** [review]
Thanks for the change!
I tried this with your change, I can see both cilium ingress and egress bpf program attached to a veth. I can connect to `kubernetes.default` from a pod as well.

One thing to note is Retina's egress bpf program is not attached to the veth. The same image attaches both the ingress and egress tc program to every veth in a non-cilium cluster.

File: pkg/plugin/packetparser/_cprog/packetparser.c [hunk 2]
```diff
@@ -247,21 +247,21 @@ int endpoint_egress_filter(struct __sk_buff *skb)
 	// So egress on host is ingress on endpoint and vice versa.
 	parse(skb, TO_ENDPOINT);
 	// Always return 0 to allow packet to pass.
```
Comment (anubhabMajumdar) [inline pkg/plugin/packetparser/_cprog/packetparser.c:249]: Can we update the comment as well?

**Review by anubhabMajumdar — APPROVED** [review]

**Comment by pchaigno** [comment]
@microsoft-github-policy-service agree

File: pkg/plugin/packetparser/_cprog/packetparser.c (hunk 1 above)
Comment (pchaigno) [reply pkg/plugin/packetparser/_cprog/packetparser.c]: Retina doesn't seem to assume kernel headers are installed so I followed the approached taken for e.g. `ETH_P_IP` and copied the definition of `TC_ACT_UNSPECIFIED` into `packetparser.h`.

**Comment by pchaigno** [comment]
> One thing to note is Retina's egress bpf program is not attached to the veth. The same image attaches both the ingress and egress tc program to every veth in a non-cilium cluster.

I suspect that might be because Cilium wasn't installed with `--bpf-filter-priority=2`. I tested this patch on GKE with Cilium installed with the following command and was able to run the full Cilium connectivity tests successfully.
```
helm install cilium cilium/cilium \
	--version 1.15.3 \
	--namespace kube-system \
	--set nodeinit.enabled=true \
	--set nodeinit.reconfigureKubelet=true \
	--set nodeinit.removeCbrBridge=true \
	--set cni.binPath=/home/kubernetes/bin \
	--set gke.enabled=true \
	--set ipam.mode=kubernetes \
	--set ipv4NativeRoutingCIDR=$NATIVE_CIDR \
	--set-string extraConfig.bpf-filter-priority=2 \
	--set debug.enabled=true
```
Retina was also attached to both ingress and egress, alongside Cilium:
```
$ ks exec ds/cilium -c cilium-agent -- bpftool net | grep lxc461b2341ecd6
lxc461b2341ecd6(18) clsact/ingress endpoint_ingres id 3283
lxc461b2341ecd6(18) clsact/ingress cil_from_container-lxc461b2341ecd6 id 790
lxc461b2341ecd6(18) clsact/egress endpoint_egress id 3282
lxc461b2341ecd6(18) clsact/egress cil_to_container-lxc461b2341ecd6 id 795
```

**Comment by rbtr** [comment]
@pchaigno would you mind signing your commit here? I'll be able to merge it in then 🙂 
Thanks for the fix ❤️

**Review by tamilmani1989 — APPROVED** [review]
lgtm. Thanks for this fix @pchaigno

**Comment by pchaigno** [comment]
> @pchaigno would you mind signing your commit here? I'll be able to merge it in then 🙂

My bad. It seems I had read the contribution guidelines a bit too quickly. It should be fixed now.

**Pull request merged**